src/futils.cpp

```diff
@@ -139,123 +139,89 @@ namespace Exiv2 {
+                uint32_t octet_a = i < dataLength ? data[i++] : 0 ;
+                uint32_t octet_b = i < dataLength ? data[i++] : 0 ;
+                uint32_t octet_c = i < dataLength ? data[i++] : 0 ;
+
+                uint32_t triple = (octet_a << 0x10) + (octet_b << 0x08) + octet_c;
+
+                result[j++] = encoding_table[(triple >> 3 * 6) & 0x3F];
+                result[j++] = encoding_table[(triple >> 2 * 6) & 0x3F];
+                result[j++] = encoding_table[(triple >> 1 * 6) & 0x3F];
+                result[j++] = encoding_table[(triple >> 0 * 6) & 0x3F];
             }
 
-            /*
-            * if we have all three bytes available, then their encoding is spread
-            * out over four characters
-            */
-            if((x+2) < dataLength)
-            {
-                if(resultIndex >= resultSize) return 0;   /* indicate failure: buffer too small */
-                result[resultIndex++] = base64chars[n3];
-            }
+            for (size_t i = 0; i < mod_table[dataLength % 3]; i++)
+                result[output_length - 1 - i] = '=';
+            result[output_length]=0;
         }
-
-        /*
-        * create and add padding that is required if we did not have a multiple of 3
-        * number of characters available
-        */
-        if (padCount > 0)
-        {
-            for (; padCount < 3; padCount++)
-            {
-                if(resultIndex >= resultSize) return 0;   /* indicate failure: buffer too small */
-                result[resultIndex++] = '=';
-            }
-        }
-        if(resultIndex >= resultSize) return 0;   /* indicate failure: buffer too small */
-        result[resultIndex] = 0;
-        return 1;   /* indicate success */
+        return rc;
     } // base64encode
 
-    long base64decode(const char *in, char *out, size_t out_size) {
-        static const char decode[] = "|$$$}rstuvwxyz{$$$$$$$>?@ABCDEFGHIJKLMNOPQRSTUVW"
-                         "$$$$$$XYZ[\\]^_`abcdefghijklmnopq";
-        long len;
-        long i;
-        long done = 0;
-        unsigned char v;
-        unsigned char quad[4];
-
-        while (*in) {
-            len = 0;
-            for (i = 0; i < 4 && *in; i++) {
-                v = 0;
-                while (*in && !v) {
-                    v = *in++;
-                    v = (v < 43 || v > 122) ? 0 : decode[v - 43];
-                    if (v)
-                        v = (v == '$') ? 0 : v - 61;
-                    if (*in) {
-                        len++;
-                        if (v)
-                            quad[i] = v - 1;
-                    } else
-                        quad[i] = 0;
-                }
+    long base64decode(const char* in,char* out,size_t out_size) {
+        long   result       = 0;
+        size_t input_length = in ? ::strlen(in) : 0;
+        if (!in || input_length % 4 != 0) return result;
+        
+        unsigned char encoding_table[]= {'A', 'B', 'C', 'D', 'E', 'F', 'G', 'H',
```

Comment:
I think it would be nice if w can move this table out of this function and share it with the `encode` function, no?

Reply:
Yes.  I thought of that.  In decode it's a `char**` and in encode it's `unsigned char**` (or the other way).  And I thought "what happens if this is exposed as an external in the library?".   So, I thought _leave it alone_ - it's working.

Reply:
Since we are in a `.cpp` we can just put it into an anonymous namespace and wouldn't have to worry about leaking it.